webapp/sources/rudder/rudder-core/src/main/scala/com/normation/cfclerk/domain/TechniqueVersion.scala

```diff
@@ -37,290 +37,65 @@
- * the upstream_version if epoch is equal. epoch is compared numerically. The
- * upstream_version part is compared using the following algorithm :
- *
- * The strings are compared from left to right.
- *
- * First the initial part of each string consisting entirely of non-digit
- * characters is determined. These two parts (one of which may be empty) are
- * compared lexically. If a difference is found it is returned. The lexical
- * comparison is a comparison of ASCII values modified so that all the letters sort
- * earlier than all the non-letters and so that a tilde sorts before anything, even
- * the end of a part. For example, the following parts are in sorted order from
- * earliest to latest: ~~, ~~a, ~, the empty part, a.
- *
- * Then the initial part of the remainder of each string which consists entirely of
- * digit characters is determined. The numerical values of these two parts are
- * compared, and any difference found is returned as the result of the
- * comparison. For these purposes an empty string (which can only occur at the end
- * of one or both version strings being compared) counts as zero.
- *
- * These two steps (comparing and removing initial non-digit strings and initial
- * digit strings) are repeated until a difference is found or both strings are
- * exhausted.
- */
+import com.normation.GitVersion.RevId
+import com.normation.utils.ParseVersion
+import com.normation.utils.Version
+
 
 class TechniqueVersionFormatException(msg: String) extends Exception("The version format of a technique should be : [epoch:]upstream_version\n" + msg)
 
-final class TechniqueVersion(val epoch: Int, val upsreamTechniqueVersion: UpstreamTechniqueVersion) extends Ordered[TechniqueVersion] {
-  def compare(v: TechniqueVersion): Int = {
-    if (epoch != v.epoch) epoch compare v.epoch
-    else upsreamTechniqueVersion compare v.upsreamTechniqueVersion
+final case class TechniqueVersion(version: Version, revId: Option[RevId] = None) extends Ordered[TechniqueVersion] {
+  def compare(v: TechniqueVersion): Int = (version compare v.version) match {
+    case 0 =>
+      // we can't compare two revId, so just use alpha-num order
+      // A revId is always before none (which means "HEAD")
+      (revId, v.revId) match {
+        case (None   , None   ) =>  0
+        case (None   , _      ) =>  1
+        case (_      , None   ) => -1
+        case (Some(x), Some(y)) =>  String.CASE_INSENSITIVE_ORDER.compare(x.value, y.value)
+      }
+    case i => i
   }
 
-  override def equals(other:Any) : Boolean = other match {
-    case that:TechniqueVersion => this.epoch == that.epoch && this.upsreamTechniqueVersion == that.upsreamTechniqueVersion
-    case _ => false
-  }
+  def withDefaultRevId = TechniqueVersion(version)
 
-  override lazy val hashCode : Int = 7 + 13 * epoch + 41 * upsreamTechniqueVersion.hashCode
+  // intended for debug
+  def debugString = serialize
+  // for serialisation on path
+  def serialize = revId.fold(
```

Comment:
why a fold an not just a pattern matching ?

or even a version.toVersionString + revId.map(_.value).getOrElse("") ?

Reply:
not sure, most of the time I find fold clearer (not on `Option`, though, which is a type inference horror), so I default to it. I can change it.